File: src/gpytoolbox/particle_swarm.py
import numpy as np

def particle_swarm(fun,lb,ub,n_particles=100,max_iter=100,momentum=0.9,phi=0.1,verbose=False, topology='full'):
    """Particle swarm optimization.
    
    Parameters
    ----------
    fun : callable
        Function to minimize
    lb : (n,) numpy double array
        Vector of lower bounds
    ub : (n,) numpy double array
        Vector of upper bounds
    n_particles : int, optional (default: 100)
        Number of particles
    max_iter : int, optional (default: 1000)
        Maximum number of iterations
    topology : str, optional (default: 'full')
        Connectivity of the swarm. Either 'full' or 'ring'.
    verbose : bool, optional (default: False)
        Print progress to stdout
    
    Returns
    -------
    x : (n,) numpy double array
        Best solution
    f : double
        Best objective value
    """
    current_best_f = np.inf
    n = len(lb)
    x = np.random.uniform(lb,ub,(n_particles,n))
    best_xi = x.copy()
    best_fi = np.inf*np.ones(n_particles)
    for i in range(n_particles):
        # This x
        xi = x[i,:]
        # print(xi)
        f = fun(xi)
        # print(xi)
        best_xi[i,:] = xi.copy()
        best_fi[i] = np.squeeze(f.copy())
        # if verbose:
            # print("Particle %d: f = %f" % (i,f))
        if f < current_best_f:
            current_best_x = xi.copy()
            current_best_f = f.copy()
    # print(current_best_x.reshape((-1, 3)))
    # initialize particle velocities
    velocity_lb = lb - ub
    velocity_ub = ub - lb
    v = np.random.uniform(velocity_lb,velocity_ub,(n_particles,n))

    # Repeat until convergence
    for iter in range(max_iter):
        for i in range(n_particles):
            # Pick random numbers rp and rg between 0 and 1
            rp = np.random.uniform()
            rg = np.random.uniform()
            # Update velocity
            if topology == 'full':
                best_neighbor = current_best_x.copy()
            elif topology == 'ring':
                # Pick the best among the current particle and its two neighbors
                neighbors = np.array([i-1,i,i+1])%n_particles
                xi_neighbors = np.array([best_xi[j,:] for j in neighbors])
                fi_neighbors = np.array([best_fi[j] for j in neighbors])
                best_neighbor = xi_neighbors[np.argmin(fi_neighbors)]
            v[i,:] = momentum*v[i,:] + phi*rp*(best_xi[i,:] - x[i,:]) + phi*rg*(best_neighbor - x[i,:])
            # Update position
            x[i,:] = x[i,:] + v[i,:]
            # Check bounds
            x[i,:] = np.maximum(x[i,:],lb)
            x[i,:] = np.minimum(x[i,:],ub)
            # Evaluate objective
            f = fun(x[i,:])
            # Update best position
            if f < best_fi[i]:
                best_xi[i,:] = x[i,:].copy()
                best_fi[i] = np.squeeze(f.copy())
                if f < current_best_f:
                    current_best_x = x[i,:].copy()
                    current_best_f = f.copy()
        if verbose:
            print("Iteration %d: f = %f" % (iter,current_best_f))
    # print(current_best_x)
    return current_best_x, current_best_f
    

File: src/cpp/particle_swarm.cpp
#include "particle_swarm.h"
#include <random>
#include <limits>
#include <iostream>
#include <algorithm>

void particle_swarm(
    const std::function<double(const Eigen::VectorXd&)>& fun,
    const Eigen::VectorXd& lb,
    const Eigen::VectorXd& ub,
    int n_particles,
    int max_iter,
    double momentum,
    double phi,
    bool verbose,
    const std::string& topology,
    Eigen::VectorXd& best_x,
    double& best_f)
{
    const int n = lb.size();
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_real_distribution<double> uniform01(0.0, 1.0);

    // Initialize particle positions uniformly in [lb, ub].
    Eigen::MatrixXd x(n_particles, n);
    for (int i = 0; i < n_particles; ++i) {
        for (int j = 0; j < n; ++j) {
            x(i, j) = lb(j) + uniform01(gen) * (ub(j) - lb(j));
        }
    }

    Eigen::MatrixXd best_xi = x;
    Eigen::VectorXd best_fi = Eigen::VectorXd::Constant(
        n_particles, std::numeric_limits<double>::infinity());

    double current_best_f = std::numeric_limits<double>::infinity();
    Eigen::VectorXd current_best_x = x.row(0).transpose();

    for (int i = 0; i < n_particles; ++i) {
        Eigen::VectorXd xi = x.row(i).transpose();
        double f = fun(xi);
        best_xi.row(i) = xi.transpose();
        best_fi(i) = f;
        if (f < current_best_f) {
            current_best_x = xi;
            current_best_f = f;
        }
    }

    // Initialize velocities uniformly in [lb-ub, ub-lb].
    Eigen::MatrixXd v(n_particles, n);
    for (int i = 0; i < n_particles; ++i) {
        for (int j = 0; j < n; ++j) {
            const double vlb = lb(j) - ub(j);
            const double vub = ub(j) - lb(j);
            v(i, j) = vlb + uniform01(gen) * (vub - vlb);
        }
    }

    for (int iter = 0; iter < max_iter; ++iter) {
        for (int i = 0; i < n_particles; ++i) {
            const double rp = uniform01(gen);
            const double rg = uniform01(gen);

            Eigen::VectorXd best_neighbor;
            if (topology == "full") {
                best_neighbor = current_best_x;
            } else if (topology == "ring") {
                const int prev = (i - 1 + n_particles) % n_particles;
                const int next = (i + 1) % n_particles;
                int best_idx = i;
                double best_val = best_fi(i);
                if (best_fi(prev) < best_val) {
                    best_val = best_fi(prev);
                    best_idx = prev;
                }
                if (best_fi(next) < best_val) {
                    best_val = best_fi(next);
                    best_idx = next;
                }
                best_neighbor = best_xi.row(best_idx).transpose();
            } else {
                best_neighbor = current_best_x;
            }

            for (int j = 0; j < n; ++j) {
                v(i, j) = momentum * v(i, j)
                    + phi * rp * (best_xi(i, j) - x(i, j))
                    + phi * rg * (best_neighbor(j) - x(i, j));
                x(i, j) = x(i, j) + v(i, j);
                x(i, j) = std::max(x(i, j), lb(j));
                x(i, j) = std::min(x(i, j), ub(j));
            }

            Eigen::VectorXd xi = x.row(i).transpose();
            const double f = fun(xi);
            if (f < best_fi(i)) {
                best_xi.row(i) = xi.transpose();
                best_fi(i) = f;
                if (f < current_best_f) {
                    current_best_x = xi;
                    current_best_f = f;
                }
            }
        }
        if (verbose) {
            std::cout << "Iteration " << iter << ": f = " << current_best_f << std::endl;
        }
    }

    best_x = current_best_x;
    best_f = current_best_f;
}
